Support component customisation (#709)
withastro/starlight: docs/src/content/docs/guides/overriding-components.md

## How to override

1. Choose the Starlight component you want to override.
   You can find a full list of components in the [Overrides Reference](/reference/overrides/)

   This example will override Starlight’s [`SocialIcons`](/reference/overrides/#socialicons) component in the page nav bar.

2. Create an Astro component to replace the Starlight component with.
   This example renders a contact link.

   ```astro
   ---
   // src/components/EmailLink.astro
   import type { Props } from '@astrojs/starlight/props';
   ---

   <a href="mailto:houston@example.com">E-mail Me</a>
   ```

3. Tell Starlight to use your custom implementation in the [`components`](/reference/configuration/#components) configuration option in `astro.config.mjs`:

   ```js {9-12}
   // astro.config.mjs
   import { defineConfig } from 'astro/config';
   import starlight from '@astrojs/starlight';

   export default defineConfig({
     integrations: [
       starlight({
         title: 'My Docs with Overrides',
         components: {
           // Override the default `SocialLinks` component.
           SocialIcons: './src/components/EmailLink.astro',
         },
       }),
     ],
   });
   ```

> ## Extend a built-in component

Author: ```suggestion
## Reuse a built-in component
```
Alternative? Would love to avoid a super long heading if possible.

You can add some UI to your site without completely replacing Starlight’s default UI by extending a built-in component.

> To extend a built-in component, import and render it in your custom component. In this example, we render a custom e-mail link and then also render the default `SocialLinks` component:

Author: I like this!

```astro {4,8}
---
// src/components/EmailLink.astro
import type { Props } from '@astrojs/starlight/props';
import Default from '@astrojs/starlight/SocialIcons.astro';
---

<a href="mailto:houston@example.com">E-mail Me</a>
<Default {...Astro.props}><slot /></Default>
```

When rendering a built-in component:

- Spread `Astro.props` into it. This makes sure that it receives all the data it needs to render.
- Add a `<slot />` inside the component. This makes sure that if the component is passed any child elements, Astro knows where to render them.

## Use page data

> When overriding a Starlight component, your custom implementation receives a standard `Astro.props` object.
> This object contains all the data for the current page to help you decide what to render.

Author: Maybe this? You technically _could_ “set” one of these values (e.g. intercept values and change them before passing them to the built-in Starlight component), but I don’t think that’s a common use case or what we’re talking about here.

```suggestion
When overriding a Starlight component, your custom implementation receives a standard `Astro.props` object with all the data for the current page.
This allows you to use these values to control how your component template renders.
```

For example, you can read the page’s frontmatter values as `Astro.props.entry.data`. In the following example, a replacement [`PageTitle`](/reference/overrides/#pagetitle) component uses this to display the current page’s title:

```astro {5} "{title}"
---
// src/components/Title.astro
import type { Props } from '@astrojs/starlight/props';

const { title } = Astro.props.entry.data;
---

<h1 id="_top">{title}</h1>

<style>
  h1 {
    font-family: 'Comic Sans';
  }
</style>
```

Learn more about all the available props in the [Overrides Reference](/reference/overrides/#prop-types).

> ### Only override on specific pages
> 
> Component overrides apply to all pages. However, you can use `Astro.props` to determine when to show your custom UI and when to show Starlight’s default UI.
> 
> In the following example, a component overriding the [`Footer`](/reference/overrides/#footer-1) renders a custom footer on the homepage and shows the default footer on all other pages:

Author: > when I read the heading re only overriding on specific pages, I expected to see some kind of configuration option, not just conditional rendering

This was kind of the intention. One thing I anticipate is exactly this question: “OK cool, I want to to override, but only on page X” or “only on pages A, B, C”. So this was aiming to capture that user intent and then explain how to do it.

I do like the copy changes here though! Do you think it would be OK to keep the current heading and just use those? Or do you think the heading is too sneaky? 😅 

```suggestion
### Only override on specific pages

Component overrides apply to all pages. However, you can conditionally render using values from `Astro.props` to determine when to show your custom UI,  when to show Starlight’s default UI, or even when to show something entirely different.

In the following example, a component overriding Starlight's [`Footer`](/reference/overrides/#footer-1) displays "Built with Starlight 🌟" on the homepage only, and otherwise shows the default footer on all other pages:
```

> In the following example, a component overriding the [`Footer`](/reference/overrides/#footer-1) renders a custom footer on the homepage and shows the default footer on all other pages:

Reviewer: Yes would be fine! (I like front-loading the conditional rendering because again I was asking myself, "Wait, what am I missing? Isn't this just conditional rendering?" So I get that, and you get your sneaky headline!)

---
title: Overriding Components
description: Learn how to override Starlight’s built-in components to add custom elements to your documentation site’s UI.
sidebar:
  badge: New
---

> Starlight’s default UI and configuration options are designed to be flexible and work for a range of content. Much of [Starlight's default appearance can be customized](/guides/customization/) with CSS and configuration options.

Author: Separate CSS & Customization links:

```suggestion
Starlight’s default UI and configuration options are designed to be flexible and work for a range of content. Much of Starlight's default appearance can be customized with [CSS](/guides/css-and-tailwind/) and [configuration options](/guides/customization/).
```

When you need more than what’s possible out of the box, Starlight supports building your own custom components to extend or override (completely replace) its default components.

## When to override

Overriding Starlight’s default components can be useful when:

- You want to change how a part of Starlight’s UI looks in a way not possible with [custom CSS](/css-and-tailwind/).
- You want to change how a part of Starlight’s UI behaves.
- You want to add some additional UI alongside Starlight’s existing UI.

## How to override

1. Choose the Starlight component you want to override.

>    You can find a full list of components in the [Overrides Reference](/reference/overrides/)

Author: Missing full stop:

```suggestion
   You can find a full list of components in the [Overrides Reference](/reference/overrides/).
```

   import starlight from '@astrojs/starlight';

   export default defineConfig({
     integrations: [
       starlight({
         title: 'My Docs with Overrides',
         components: {
           // Override the default `SocialLinks` component.
           SocialIcons: './src/components/EmailLink.astro',
         },
       }),
     ],
   });
   ```

## Extend a built-in component

You can build with Starlight’s default UI components just as you would with your own: importing and rendering them in your own custom components. This allows you to keep all of Starlight's basic UI within your design, while adding extra UI alongside them.

The example below shows a custom component that renders an e-mail link along with the default `SocialLinks` component:

```astro {4,8}
---
// src/components/EmailLink.astro
import type { Props } from '@astrojs/starlight/props';
import Default from '@astrojs/starlight/SocialIcons.astro';
---

<a href="mailto:houston@example.com">E-mail Me</a>
<Default {...Astro.props}><slot /></Default>
```

When rendering a built-in component inside a custom component:

- Spread `Astro.props` into it. This makes sure that it receives all the data it needs to render.

> - Add a `<slot />` inside the default component. This makes sure that if the component is passed any child elements, Astro knows where to render them.

Author: Astro docs link:

```suggestion
- Add a [`<slot />`](https://docs.astro.build/en/core-concepts/astro-components/#slots) inside the default component. This makes sure that if the component is passed any child elements, Astro knows where to render them.
```
